packages/wast-parser/src/tokenizer.js

```diff
@@ -293,26 +300,37 @@ function tokenize(input: string) {
 
       let numberLiterals = NUMBERS;
 
-      if (char === "0" && lookahead().toUpperCase() === "X") {
+      if (char === "0" && lookahead() === "x") {
         value += "0x";
         numberLiterals = HEX_NUMBERS;
         eatCharacter(2);
       }
 
       while (
-        numberLiterals.test(char) ||
+        (char !== undefined && numberLiterals.test(char)) ||
         (lookbehind() === "p" && char === "+") ||
-        (lookbehind().toUpperCase() === "E" && char === "-") ||
-        (value.length > 0 && char.toUpperCase() === "E")
+        (lookbehind() === "p" && char === "-") ||
+        (lookbehind() === "e" && char === "+") ||
+        (lookbehind() === "e" && char === "-") ||
+        (value.length > 0 && (char === "e" || char === "E"))
       ) {
         if (char === "p" && value.includes("p")) {
-          throw new Error("Unexpected character `p`.");
+          throw new Error('Unexpected character "p"');
```

Comment:
We can maybe have a magic helper for this, like `throwUnexpectedChar` which would take the char with a closure.